### DISCUSSION TITLE
Chunk-Store encryption capabilities

### DISCUSSION THREAD

**Discussion by robert-vcs** [rauc/rauc#1480] [Q&A] [answered]
Hello everyone,

I'm currently exploring the encryption capabilities in RAUC for adaptive updates, specifically focusing on the chunk-store. I've read the documentation and sections about bundle encryption capabilities but was not sure if it is also applicable to adaptive RAUC updates. As far as I understand, that's limited to bundles (full or delta) composed by the RAUC server/host. 

Additionally to that, can anyone comment if there is a feature available to encrypt the chunks within in a chunk-store? The described features I've found are targeting data in transit (secured by TLS) only. In case encryption of chunks is not a supported capability, what is the rational of limiting the encryption to bundle scenarios? Might this be a future improvement or is there a conscious decision to omit that available?

I would greatly appreciate any guidance, documentation references, or insights from the community regarding the supported encryption mechanisms and update processes within adaptive RAUC.

Thank you for your assistance!

**Comment by jluebbe** [comment, marked as answer]
> I'm currently exploring the encryption capabilities in RAUC for adaptive updates, specifically focusing on the chunk-store. I've read the documentation and sections about bundle encryption capabilities but was not sure if it is also applicable to adaptive RAUC updates. As far as I understand, that's limited to bundles (full or delta) composed by the RAUC server/host.
>
> Additionally to that, can anyone comment if there is a feature available to encrypt the chunks within in a chunk-store? The described features I've found are targeting data in transit (secured by TLS) only.

To clarify, there are two completely separate ways RAUC supports "chunk-based" updates: casync and adaptive streaming updates. casync has the only index in the bundle and a separate chunk store (which is not encrypted at rest).

Adaptive updates use streaming to download only the required parts of the bundle. For example with the block-hash-index, all the actual data is still part of the bundle and can be encrypted at rest (in addition to TLS).

For new projects, we'd suggest not using casync any more.

> In case encryption of chunks is not a supported capability, what is the rational of limiting the encryption to bundle scenarios? Might this be a future improvement or is there a conscious decision to omit that available?

For casync, it's simply a consequence of the casync tool not supporting encryption. We added streaming access to RAUC as it allows us to abstract the access to, authentication and (optionally) encryption of the bundle contents from the different ways images could be encoded. This way, we can add different adaptive update methods as needed by different use-cases and they all use authentication/encryption transparently. The same bundle file can be used locally via a USB memory stick or streamed from a HTTPS server/CDN. In the streaming case, only the actually required data (plus possibly some alignment overhead) is transferred.

@ejoerns had a presentation at FOSDEM 2023 about streaming and adaptive updates in RAUC: https://archive.fosdem.org/2023/schedule/event/delta_like_ota_streaming/
